measureme/src/counters.rs

```diff
@@ -0,0 +1,869 @@
+            let (counter, offset, pmc_width) = if cfg!(accurate_seqlock_rdpmc) && false {
+                self.access_mmap_page_with_seqlock(|mp| {
+                    let caps = unsafe { mp.__bindgen_anon_1.__bindgen_anon_1 };
+                    assert_ne!(caps.cap_user_rdpmc(), 0);
+
+                    (
+                        rdpmc(mp.index.checked_sub(1).unwrap()),
+                        mp.offset,
+                        mp.pmc_width,
+                    )
+                })
+            } else {
+                (rdpmc(self.reg_idx), 0, 48)
+            };
+
+            let counter = offset + (counter as i64);
+
+            // Sign-extend the `pmc_width`-bit value to `i64`.
+            (counter << (64 - pmc_width) >> (64 - pmc_width)) as u64
+        }
+    }
+
+    impl super::HwCounterRead for (&Counter, &Counter) {
+        type Output = (u64, u64);
+
+        #[inline]
+        fn read(&self) -> (u64, u64) {
+            // HACK(eddyb) keep the accurate code around while not using it,
+            // to minimize overhead without losing the more complex implementation.
+            if (cfg!(accurate_seqlock_rdpmc) || cfg!(unserialized_rdpmc)) && false {
+                return (self.0.read(), self.1.read());
+            }
+
+            let pmc_width = 48;
+
+            let (a_counter, b_counter) = rdpmc_pair(self.0.reg_idx, self.1.reg_idx);
+
+            // Sign-extend the `pmc_width`-bit values to `i64`.
+            (
+                ((a_counter as i64) << (64 - pmc_width) >> (64 - pmc_width)) as u64,
+                ((b_counter as i64) << (64 - pmc_width) >> (64 - pmc_width)) as u64,
+            )
+        }
+    }
+
+    /// Read the hardware performance counter indicated by `reg_idx`.
+    ///
+    /// If the counter is signed, sign extension should be performed based on
+    /// the width of the register (32 to 64 bits, e.g. 48-bit seems common).
+    #[inline(always)]
+    fn rdpmc(reg_idx: u32) -> u64 {
+        let (lo, hi): (u32, u32);
+        unsafe {
+            // NOTE(eddyb) below comment is outdated (the other branch uses `cpuid`).
+            if cfg!(unserialized_rdpmc) && false {
+                // FIXME(eddyb) the Intel and AMD manuals warn about the need for
+                // "serializing instructions" before/after `rdpmc`, if avoiding any
+                // reordering is desired, but do not agree on the full set of usable
+                // "serializing instructions" (e.g. `mfence` isn't listed in both).
+                //
+                // The only usable, and guaranteed to work, "serializing instruction"
+                // appears to be `cpuid`, but it doesn't seem easy to use, especially
+                // due to the overlap in registers with `rdpmc` itself, and it might
+                // have too high of a cost, compared to serialization benefits (if any).
+                asm!("rdpmc", in("ecx") reg_idx, out("eax") lo, out("edx") hi, options(nostack));
```

Comment:
Do we need this dead code? I think we can probably just move this comment down to the else block.

Reply:
I found it useful to have several different versions around, for quick testing (without having to dig up `git reflog` or w/e), but I'm not sure how to best control it. Maybe a trait with associated `const`s - even if it's not publicly exposed for now.